daemon: Do not attach bpf_host to L3 dev if skb_change_head is unavailable

brb opened the pull request:
See commit msgs.

brb committed:
bpf: Add ETH_HLEN=0 to stress test complexity

The PR "datapath: Support NodePort BPF on L2-less devices" [1] has
increased the complexity of bpf_host and bpf_lxc by introducing a
support for ETH_HLEN=0.

Extend the base options by adding ETH_HLEN=0 to stress test the verifier
complexity when running on net-next (ETH_HLEN=0 depends on the
skb_change_head helper which was introduced in 5.8).

[1]: https://github.com/cilium/cilium/pull/14858

Suggested-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno reviewed (commented):
Is `ETH_HLEN` only ever used in bpf_lxc (i.e., not in bpf_host)?

twpayne reviewed (approved):
LGTM

brb committed:
daemon: Ignore L3 devices on unsupported kernels

Skip devices which don't have L2 addr, if the kernel doesn't have
skb_change_head helper (the case for < 5.8 kernels).

Without the helper it is not possible to create the headroom for L2 hdr.
Unfortunately, "skb_adjust_room(skb, 14, BPF_ADJ_ROOM_MAC, 0)" cannot be
used either, as it does not set "skb->mac_header" which makes the packet
to be dropped by [1] when redirecting to another device.

[1]: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/core/filter.c?h=v5.8#n2118

Signed-off-by: Martynas Pumputis <m@lambda.lt>

jibi commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -416,10 +416,14 @@ func handleNativeDevices(strict bool) {
+			log.Infof("Switching direct routing device from %s to %s", prevDev,
+				option.Config.DirectRoutingDevice)
+		} else {
+			log.Warn("NodePort BPF forwarded traffic won't be encrypted by Wireguard (requires >= 5.8 kernel)")
```
I'm not super comfortable having this as just a warning as it means that contrary to what the user expects some traffic is not going to be encrypted and the only way to find this out is by looking into the logs

borkmann replied:
Agree, shouldn't we fatal instead? It seems odd from one kernel version to another to not encrypt / encrypt.

brb replied:
Yeah, makes sense :+1:  Changed.

pchaigno commented on bpf/Makefile:
```diff
@@ -86,6 +86,8 @@ MAX_BASE_OPTIONS = -DSKIP_DEBUG=1 -DENABLE_IPV4=1 -DENABLE_IPV6=1 \
 # IPSec is incompatible with BPF NodePort so we only enable on 4.9.
 MAX_BASE_OPTIONS += -DENABLE_IPSEC=1
+else ifeq ("$(KERNEL)","netnext")
+MAX_BASE_OPTIONS += -DETH_HLEN=0
```
That should be moved below, there's already a case for netnext.

pchaigno commented on bpf/Makefile:
```diff
@@ -90,7 +90,8 @@ else
 ifeq ("$(KERNEL)","netnext")
 MAX_BASE_OPTIONS += -DENABLE_TPROXY=1 -DENABLE_REDIRECT_FAST=1 -DENABLE_BANDWIDTH_MANAGER=1 \
-	-DENABLE_HOST_SERVICES_FULL=1 # Required to work around complexity issue #14234.
+	-DENABLE_HOST_SERVICES_FULL=1 \ # Required to work around complexity issue #14234.
```
The test failure is because of this backslash followed by a whitespace. I'm having the same issue in another PR, will post here when I resolve if you haven't already.

pchaigno replied:
There's apparently no way to include inline comments inside a multiline macro definition so I ended up moving the comment before the definition. See https://github.com/cilium/cilium/pull/15540/commits/8531a11ff69dea976cc1cec6e9aa8b763e3f14e8.

pchaigno reviewed (approved):
Mostly to catch the newly-required cilium/loader review, but I also reviewed latest changes.
